geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package org.apache.geode.redis.internal;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.apache.geode.cache.execute.Function;
+import org.apache.geode.cache.execute.FunctionContext;
+import org.apache.geode.cache.execute.FunctionService;
+import org.apache.geode.cache.execute.ResultCollector;
+
+public class PubSub {
+  public static final String REDIS_PUB_SUB_FUNCTION_ID = "redisPubSubFunctionID";
+
+  Subscribers subscribers = new Subscribers();
+
+  public long publish(String channel, String message) {
+    ResultCollector<?, ?> subscriberCountCollector = FunctionService
+        .onMembers()
+        .setArguments(new String[] {channel, message})
+        .execute(REDIS_PUB_SUB_FUNCTION_ID);
+
+    List<Long> subscriberCounts = (List<Long>) subscriberCountCollector.getResult();
+    long totalSubscribers = 0;
+    for (long subscriberCount : subscriberCounts) {
+      totalSubscribers += subscriberCount;
+    }
+    return totalSubscribers;
+  }
+
+  public long subscribe(String channel, ExecutionHandlerContext context, Client client) {
+    if (subscribers.exists(channel, client)) {
+      return subscribers.findSubscribers(client).size();
+    }
```

Comment:
I'm not sure I understand. We need to know two things:
1 - has a client already subscribed to a channel
2 - how many subscriptions a client currently has

In order to address both of these concerns, the `findSubscribers` method would probably need to be inlined which seems like it would be less readable.

Reply:
You are correct. 
That said.. What is the max number of Pub/Sub combinations we can expect? How does this approach scale?
Maybe two maps <channel,List<Subscriber>> and <client,List<Subscriber>> ?

Reply:
Could we leave optimization until it's actually determined that it is needed?

Reply:
I agree, premature optimization.... But I'd be interested in what the "normal" usage on this would be. Maybe we add a load testing ticket on this, to see what load this construct can take. Is it 100's or 1000's or even 10,000's...

But I think it would be good to understand when it will slow down / break.